### PR TITLE
Daily weekly monthly yearly journal implementation

### LINKED ISSUE #59: Option to keep weekly or monthly diaries instead of just daily?
I want my journal to be formatted something like this, with a single file for each week.

```
* Monday, 01/04/16
** 10:28 Company meeting
Endless discussions about projects. Not much progress
* Tuesday, 01/05/16
** 10:30 Another company meeting
Same as yesterday.
```

I have tried adding `(setq org-journal-file-format "%Y%m%U")`. This way, on the second day onwards until the next weeks starts, new entries go to the same file (`20160101`), which is almost what I want. But this does not automatically give me a new date heading. So unless I add it manually, I end up with this:

```
* Monday, 01/04/16
** 10:28 Company meeting
Endless discussions about projects. Not much progress
** 10:30 Another company meeting
Same as yesterday.
```

I guess this is a feature request, unless you can suggest other settings to try?

### LINKED ISSUE #142: First journal entry is always at top of file above carryover items
The first journal entry of the day appears above the carryover items, and subsequent entries are below the carryover items.  The inconsistency bothers me and I can't figure out how to fix it.

I've tried inserting a (goto-char (point-max)) in the new entry function, and tried moving the new entry function under the one that inserts the carryover items.  No matter what I try, that firsrt new entry insists on being at the top of the file.  I'm sure it's a simple fix, I'm just don't know Lisp well enough, but I'm learning.

Each day I have to erase that first time stamp, go under the carryover items, and continue from t here.  It's cosmetic and not a big deal, it's just annoying.

Any ideas would be very much appreciated.

* Tuesday, 03/12/19...
  * 08:51  First new entry appears here
  * TODO from yesterday
  * STARTED from yesterday
  * WAIT from yesterday
  * 09:00 all other entries appear here

### PR DESCRIPTION
**Pull request by cslux**
Daily, weekly, monthly and yearly journal file implementation

- Fix doc-strings
- Add support for weekly, monthly and yearly journal files

New defcustom `org-journal-file-type`. Takes a symbol, either 'daily, 'weekly, 'monthly or 'yearly.

The `org-journal-file-format` is unchanged for all `org-journal-file-type`'s. But the date
is converted to the beginning of the period, see `org-journal-convert-time-to-file-type-time`
comments for an explanation.

Resolves #59
Fixes #142 
--------------------------------------------------------------------------------------------------------------------
Hey @bastibe!

I'm not 100% sure if I caught everything... Calendar support works. Carryover works.
README.org has been updated. Encryption has been tested (file and entry level
":crypt:"). Search works. Switching between `org-journal-file-type`'s is not supported
at the moment, its up to the user to transform them to 'daily, 'weekly, 'monthly or 'yearly
journal files.

I created a commit with a org-journal-test.el file for interactive development,
if you like it, merge it, if not, drop the commit.

I would suggest a new major release of org-journal (2.0.0).

@dowcet, @vkazanov, @d12frosted and @ghost  ready for testing :-)

Christian

### CODE REVIEW

**Comment by cslux** [comment]
@bastibe Ok, smashed a view more bugs, should be ready for daily use!

File: README.org [hunk 1]
```diff
@@ -14,21 +14,25 @@
...
 
-=org-journal= maintains a set of files, where each file represents a day. Convenient
-bindings allow the creation of journal records in the current daily file and search within
-all records or specified time intervals. All records can be browsed and searched from the
-Emacs Calendar for convenience. All entries in a specified TODO state will be carried over
-to the next day. Optionally, the journal can also be encrypted.
-
-An example of a daily file (it will actually look a lot nicer, depending on your org-mode
+=org-journal= maintains a set of files, where each file represents a
+day. Convenient bindings allow the creation of journal records in the current
```
Comment (bastibe) [inline README.org]: Do you want to add something here? Like, "each file represents a day (or week, month, year)".

Comment (cslux) [reply README.org]: Makes sense. I will add something.

File: org-journal.el [hunk 2]
```diff
@@ -239,15 +253,13 @@ Otherwise, date ascending."
...
 
 ;; Journal mode definition
 ;;;###autoload
-(define-derived-mode org-journal-mode org-mode "Journal"
-  "Mode for writing or viewing entries written in the journal"
+(define-derived-mode org-journal-mode org-mode
+  "Journal"
+  "Mode for writing or viewing entries written in the journal."
   (turn-on-visual-line-mode)
-  (add-hook 'after-save-hook 'org-journal-redraw-calendar nil t)
```
Comment (bastibe) [inline org-journal.el]: Why did you delete this hook?

File: org-journal.el [hunk 3]
```diff
@@ -275,12 +287,37 @@ Otherwise, date ascending."
 ;;;###autoload
 (global-set-key (kbd "C-c C-j") 'org-journal-new-entry)
 
+(defun org-journal-convert-time-to-file-type-time (&optional time)
+  "Converts time to the "
```
Comment (bastibe) [inline org-journal.el]: incomplete docstring.

File: org-journal.el [hunk 4]
```diff
@@ -327,63 +365,66 @@ Whenever a journal entry is created the
     (let* ((entry-path (org-journal-get-entry-path time))
            (should-add-entry-p (not prefix)))
 
-      ;; open journal file
+      ;; open journal fiel
       (unless (string= entry-path (buffer-file-name))
         (funcall org-journal-find-file entry-path))
+
+      (let ((entry-header
```
Comment (bastibe) [inline org-journal.el]: Could you add a short comment here to describe what this is doing?

File: org-journal.el (hunk 3 above)
Comment (cslux) [reply org-journal.el]: Whoops! Will fix that.

File: org-journal.el (hunk 4 above)
Comment (cslux) [reply org-journal.el]: Sure!

Comment (bastibe) [reply org-journal.el]: Oh, I see, this is the "empty file" block, followed by a property.

File: org-journal.el [hunk 5]
```diff
@@ -399,14 +440,15 @@ previous day's file to the current file."
                  (backward-char)
                  (setq org-map-continue-from (point))
                  subtree))))
+        (re-search-backward org-journal-created-re nil t)
```
Comment (bastibe) [inline org-journal.el]: Does this work for daily journals, which don't have a created property?

**Comment by bastibe** [comment]
Wow, thank you so, so much! This is truly exciting, and well worth a 2.0! Amazing!

File: org-journal.el (hunk 5 above)
Comment (cslux) [reply org-journal.el]: Sure! The error will be ignored (third argument -> noerror is set to `t'). But if you like I can wrap it inside a conditional!?

File: org-journal.el (hunk 4 above)
Comment (cslux) [reply org-journal.el]: Exactly!

File: org-journal.el (hunk 2 above)
Comment (cslux) [reply org-journal.el]: I tested something :-) Reverted.

**Comment by cslux** [comment]
@bastibe You're welcome! I corrected your requested changes. Ready when you are :-D

File: org-journal.el (hunk 5 above)
Comment (bastibe) [reply org-journal.el]: Ok, no need to change it. Just wanted to make sure this is not a bug.

**Comment by bastibe** [comment]
Thank you very much! I'll let this sit on master for a few days to flush out any hidden bugs, and publish a new version once we're sure it's all working.
